### PR TITLE
Shaken Fist Manifesto

### PR DESCRIPTION
**Pull request by mandoonandy**
Draft Manifesto for Shaken Fist. Trying to give guidance to newcomers about what the project is and where we want it to go.

Just a rough outline. Major edits and rewrites are insisted upon.

### CODE REVIEW

File: docs/manifesto.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+* Single machine clusters should always be possible.
+* Only the current Ubuntu LTS version is supported by the main project (forks to support other operating systems are encouraged).
+
+
+## Project Interaction Guidelines
+
+* Always polite
+* Always generous
+* Being opinionated is encouraged (but gently)
+* Updating the documentation is just important as the code change itself
```
Comment (LudvikGalois) [inline docs/manifesto.md:46]:
```suggestion
* Updating the documentation is just as important as the code change itself
```

File: docs/manifesto.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+* Avoid re-inventing the wheel (Utilise other open source projects when appropriate).
+
+
+## Feature Guidelines
+
+* Features should be deliberately limited in the options available.
+* The goal of limiting options is to reduce code complexity. If the option does not add significant code complexity then it should added.
+* The supported features and the options of those features should aim to cover the majority of use cases.
+* When a feature limits the available options, it should do so in a way that does not overly restrict a project fork from adding that option.
+* New code should conform to the conventions of the existing code base and written to be easily understood.
```
Comment (LudvikGalois) [inline docs/manifesto.md:29]:
```suggestion
* New code should conform to the conventions of the existing code base and be written to be easily understood.
```

File: docs/manifesto.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+This document attempts to list Shaken Fist's defining features, give guidance on what type of features should be added to the project, how they should be implemented and how we work together.
+
+## Shaken Fist Defining Characteristics
+
+* Shaken Fist is smaller, simpler cloud.
+* It is designed for relatively small environments with minimum management overhead.
+* It's features are highly opinionated. This means that the maintainers have chosen the best (in their opinion) features to support.
+* Opinionated features do not handle every single possible use case. This reduces the code base size thus increasing long-term maintainability.
+* The code base aims to be understandable in it's entirety by a single developer.
```
Comment (LudvikGalois) [inline docs/manifesto.md:11]:
```suggestion
* The code base is understandable in it's entirety by a single developer.
```
I don't see why this should ever be less than fact

File: docs/manifesto.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# Shaken Fist Manifesto
+
+This document attempts to list Shaken Fist's defining features, give guidance on what type of features should be added to the project, how they should be implemented and how we work together.
+
+## Shaken Fist Defining Characteristics
+
+* Shaken Fist is smaller, simpler cloud.
+* It is designed for relatively small environments with minimum management overhead.
+* It's features are highly opinionated. This means that the maintainers have chosen the best (in their opinion) features to support.
```
Comment (LudvikGalois) [inline docs/manifesto.md:9]:
```suggestion
* Its features are highly opinionated. This means that the maintainers have chosen the best (in their opinion) features to support.
```

File: docs/manifesto.md (hunk 3 above)
Comment (mandoonandy) [reply docs/manifesto.md:11]: Changed. I like your optimism.

(Some days I don't understand the code, but I may not be a representative developer.)

File: docs/manifesto.md (hunk 4 above)
Comment (mandoonandy) [reply docs/manifesto.md:9]: yep again...

File: docs/manifesto.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+* The code base is understandable in its entirety by a single developer.
+* A Shaken Fist cluster does not need a team of engineers to install or operate.
+* A Shaken Fist cluster should be simple to set up. We define 'simple' as "a person with no knowledge of the project can build a reasonable cluster in an evening".
+
+
+## Project Goals
+
+* Allow simple management of virtual machine instances without complexity.
+* Support networking between those machines and also facilitate access to external networks.
+* Avoid re-inventing the wheel (Utilise other open source projects when appropriate).
```
Comment (mikalstill) [inline docs/manifesto.md:20]: Capital U is weird here.

File: docs/index.md [hunk 6]
```diff
@@ -3,15 +3,15 @@ title: Introduction
...
 
 Thus Shaken Fist was born — named for a Simpson’s meme, Shaken Fist is a deliberately small and highly opinionated cloud implementation aimed at working well in small deployments such as homes, labs, edge compute locations, deployed systems, and so forth.
 
 I’d taken a bit of trouble with each feature in Shaken Fist to think through what the simplest and highest value way of doing something is. For example, instances always get a config drive and there is no metadata server. There is also only one supported type of virtual networking, and one supported hypervisor. That said, this means Shaken Fist is less than 5,000 lines of code, and small enough that new things can be implemented very quickly by a single middle aged developer.
 
-I think Shaken Fist is useful to others. Its apache2 licensed, and available on github if you’re interested.
+I think Shaken Fist is useful to others. Its Apache-2.0 licensed, and available on github if you’re interested.
```
Comment (mikalstill) [inline docs/index.md:14]: I feel like this entire spiel is a bit dated now. Do we blow it up? Acknowledge it as project history? Leave it?

File: docs/manifesto.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+This document attempts to list Shaken Fist's defining features, give guidance on what type of features should be added to the project, how they should be implemented and how we work together.
+
+## Shaken Fist Defining Characteristics
+
+* Shaken Fist is smaller, simpler cloud.
+* It is designed for relatively small environments with minimum management overhead.
+* Its features are highly opinionated. This means that the maintainers have chosen the best (in their opinion) features to support.
+* Opinionated features do not handle every single possible use case. This reduces the code base size thus increasing long-term maintainability.
+* The code base is understandable in its entirety by a single developer.
```
Comment (mikalstill) [inline docs/manifesto.md:11]: Specifically, I have an emotional barrier at 5,000 lines of python for the core.

**Review by mikalstill — COMMENTED** [review]
This is really good. My comments are minor.

File: docs/index.md (hunk 6 above)
Comment (mandoonandy) [reply docs/index.md:14]: I don't mind it but happy if you want to change/reorganise it.

File: docs/manifesto.md (hunk 7 above)
Comment (mandoonandy) [reply docs/manifesto.md:11]: Hmmmm
```
and@baldr:~/sf/shakenfist/shakenfist (master)$ cloc  .
      42 text files.
      42 unique files.                              
       8 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.14 s (273.6 files/s, 65162.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          37           1572            356           6884
-------------------------------------------------------------------------------
SUM:                            37           1572            356           6884
```

Comment (LudvikGalois) [reply docs/manifesto.md:11]: Is that including tests?

Comment (mandoonandy) [reply docs/manifesto.md:11]: sshhhh... Looking only for core code this time.
```
and@baldr:~/sf/shakenfist/shakenfist (master)$ cloc  --exclude-dir=tests,client .
      24 text files.
      24 unique files.                              
       2 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.03 s (934.9 files/s, 225615.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          24           1174            318           4300
-------------------------------------------------------------------------------
SUM:                            24           1174            318           4300
-------------------------------------------------------------------------------
```
 (Maybe we should break it up into a few different repositories?? :sunglasses:  )

File: docs/index.md (hunk 6 above)
Comment (mikalstill) [reply docs/index.md:14]: Maybe that's a separate change...

File: docs/manifesto.md (hunk 7 above)
Comment (mikalstill) [reply docs/manifesto.md:11]: Yeah, I exclude tests. We'll one day exceed that, but for now it makes me happy.

**Review by mikalstill — APPROVED** [review]
LGTM, thanks.

**Pull request merged**